File: sknetwork/topology/structure.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created in July 2019
@author: Nathan de Lara <nathan.delara@polytechnique.org>
@author: Quentin Lutz <qlutz@enst.fr>
@author: Thomas Bonald <tbonald@enst.fr>
"""
from typing import Tuple, Optional, Union, List

import numpy as np
from scipy import sparse

from sknetwork.utils.check import is_symmetric, check_format
from sknetwork.utils.format import get_adjacency
from sknetwork.path import get_distances


def get_connected_components(input_matrix: sparse.csr_matrix, connection: str = 'weak', force_bipartite: bool = False) \
        -> np.ndarray:
    """Extract the connected components of a graph.

    Parameters
    ----------
    input_matrix :
        Input matrix (either the adjacency matrix or the biadjacency matrix of the graph).
    connection :
        Must be ``'weak'`` (default) or ``'strong'``. The type of connection to use for directed graphs.
    force_bipartite : bool
        If ``True``, consider the input matrix as the biadjacency matrix of a bipartite graph.

    Returns
    -------
    labels :
        Connected component of each node.
        For bipartite graphs, rows and columns are concatenated (rows first).

    Example
    -------
    >>> from sknetwork.topology import get_connected_components
    >>> from sknetwork.data import house
    >>> get_connected_components(house())
    array([0, 0, 0, 0, 0], dtype=int32)
    """
    input_matrix = check_format(input_matrix)
    if len(input_matrix.data) == 0:
        raise ValueError('The graph is empty (no edge).')
    adjacency, _ = get_adjacency(input_matrix, force_bipartite=force_bipartite)
    labels = sparse.csgraph.connected_components(adjacency, connection=connection, return_labels=True)[1]
    return labels


def is_connected(input_matrix: sparse.csr_matrix, connection: str = 'weak', force_bipartite: bool = False) -> bool:
    """Check whether the graph is connected.

    Parameters
    ----------
    input_matrix :
        Input matrix (either the adjacency matrix or the biadjacency matrix of the graph).
    connection :
        Must be ``'weak'`` (default) or ``'strong'``. The type of connection to use for directed graphs.
    force_bipartite : bool
        If ``True``, consider the input matrix as the biadjacency matrix of a bipartite graph.

    Example
    -------
    >>> from sknetwork.topology import is_connected
    >>> from sknetwork.data import house
    >>> is_connected(house())
    True
    """
    return len(set(get_connected_components(input_matrix, connection, force_bipartite))) == 1


def get_largest_connected_component(input_matrix: sparse.csr_matrix, connection: str = "weak",
                                    force_bipartite: bool = False, return_index: bool = False) \
        -> Union[sparse.csr_matrix, Tuple[sparse.csr_matrix, np.ndarray]]:
    """Extract the largest connected component of a graph. Bipartite graphs are treated as undirected.

    Parameters
    ----------
    input_matrix :
        Adjacency matrix or biadjacency matrix of the graph.
    connection :
        Must be ``'weak'`` (default) or ``'strong'``. The type of connection to use for directed graphs.
    force_bipartite : bool
        If ``True``, consider the input matrix as the biadjacency matrix of a bipartite graph.
    return_index : bool
        Whether to return the index of the nodes of the largest connected component in the original graph.

    Returns
    -------
    output_matrix : sparse.csr_matrix
        Adjacency matrix or biadjacency matrix of the largest connected component.
    index : array
        Indices of the nodes in the original graph.
        For bipartite graphs, rows and columns are concatenated (rows first).

    Example
    -------
    >>> from sknetwork.topology import get_largest_connected_component
    >>> from sknetwork.data import house
    >>> get_largest_connected_component(house()).shape
    (5, 5)
    """
    input_matrix = check_format(input_matrix)
    adjacency, bipartite = get_adjacency(input_matrix, force_bipartite=force_bipartite)
    labels = get_connected_components(adjacency, connection=connection)
    unique_labels, counts = np.unique(labels, return_counts=True)
    largest_component_label = unique_labels[np.argmax(counts)]

    if bipartite:
        n_row, n_col = input_matrix.shape
        index_row = np.argwhere(labels[:n_row] == largest_component_label).ravel()
        index_col = np.argwhere(labels[n_row:] == largest_component_label).ravel()
        index = np.hstack((index_row, index_col))
        output_matrix = input_matrix[index_row, :]
        output_matrix = (output_matrix.tocsc()[:, index_col]).tocsr()
    else:
        index = np.argwhere(labels == largest_component_label).ravel()
        output_matrix = input_matrix[index, :]
        output_matrix = (output_matrix.tocsc()[:, index]).tocsr()
    if return_index:
        return output_matrix, index
    else:
        return output_matrix


def is_bipartite(adjacency: sparse.csr_matrix, return_biadjacency: bool = False) \
        -> Union[bool, Tuple[bool, Optional[sparse.csr_matrix], Optional[np.ndarray], Optional[np.ndarray]]]:
    """Check whether a graph is bipartite.

    Parameters
    ----------
    adjacency :
       Adjacency matrix of the graph (symmetric).
    return_biadjacency :
        If ``True``, return a biadjacency matrix of the graph if bipartite.

    Returns
    -------
    is_bipartite : bool
        A boolean denoting if the graph is bipartite.
    biadjacency : sparse.csr_matrix
        A biadjacency matrix of the graph if bipartite (optional).
    rows : np.ndarray
        Index of rows in the original graph (optional).
    cols : np.ndarray
        Index of columns in the original graph (optional).

    Example
    -------
    >>> from sknetwork.topology import is_bipartite
    >>> from sknetwork.data import cyclic_graph
    >>> is_bipartite(cyclic_graph(4))
    True
    >>> is_bipartite(cyclic_graph(3))
    False
    """
    if not is_symmetric(adjacency):
        raise ValueError('The graph must be undirected.')
    if adjacency.diagonal().any():
        if return_biadjacency:
            return False, None, None, None
        else:
            return False
    n = adjacency.indptr.shape[0] - 1
    coloring = np.full(n, -1, dtype=int)
    exists_remaining = n
    while exists_remaining:
        src = np.argwhere(coloring == -1)[0, 0]
        next_nodes = [src]
        coloring[src] = 0
        exists_remaining -= 1
        while next_nodes:
            node = next_nodes.pop()
            for neighbor in adjacency.indices[adjacency.indptr[node]:adjacency.indptr[node + 1]]:
                if coloring[neighbor] == -1:
                    coloring[neighbor] = 1 - coloring[node]
                    next_nodes.append(neighbor)
                    exists_remaining -= 1
                elif coloring[neighbor] == coloring[node]:
                    if return_biadjacency:
                        return False, None, None, None
                    else:
                        return False
    if return_biadjacency:
        rows = np.argwhere(coloring == 0).ravel()
        cols = np.argwhere(coloring == 1).ravel()
        return True, adjacency[rows, :][:, cols], rows, cols
    else:
        return True

File: sknetwork-rs/src/topology/structure.rs
use std::collections::VecDeque;

use sprs::{CsMat, TriMat};

use crate::utils::check::is_symmetric;
use crate::utils::format::{MatrixInput, get_adjacency};

#[derive(Debug, Clone, PartialEq, Eq)]
/// Errors raised by structure error operations.
pub enum StructureError {
    /// Indicates empty graph.
    EmptyGraph,
    /// Indicates invalid connection.
    InvalidConnection,
    /// Indicates not undirected.
    NotUndirected,
}

#[derive(Debug, Clone, PartialEq)]
/// BipartiteResult enum.
pub enum BipartiteResult {
    /// Indicates bool.
    Bool(bool),
    /// Returns bipartite status plus optional biadjacency and index partitions.
    Full(
        bool,
        Option<CsMat<f64>>,
        Option<Vec<usize>>,
        Option<Vec<usize>>,
    ),
}

fn weak_components(adjacency: &CsMat<f64>) -> Vec<usize> {
    let n = adjacency.rows();
    let undirected = adjacency + &adjacency.transpose_view().to_owned();
    let mut labels = vec![usize::MAX; n];
    let mut cc = 0usize;
    for start in 0..n {
        if labels[start] != usize::MAX {
            continue;
        }
        let mut q = VecDeque::new();
        q.push_back(start);
        labels[start] = cc;
        while let Some(u) = q.pop_front() {
            if let Some(row) = undirected.outer_view(u) {
                for &v in row.indices() {
                    if labels[v] == usize::MAX {
                        labels[v] = cc;
                        q.push_back(v);
                    }
                }
            }
        }
        cc += 1;
    }
    labels
}

fn dfs_order(graph: &CsMat<f64>, start: usize, seen: &mut [bool], order: &mut Vec<usize>) {
    let mut stack = vec![(start, false)];
    while let Some((u, processed)) = stack.pop() {
        if processed {
            order.push(u);
            continue;
        }
        if seen[u] {
            continue;
        }
        seen[u] = true;
        stack.push((u, true));
        if let Some(row) = graph.outer_view(u) {
            for &v in row.indices() {
                if !seen[v] {
                    stack.push((v, false));
                }
            }
        }
    }
}

fn strong_components(adjacency: &CsMat<f64>) -> Vec<usize> {
    let n = adjacency.rows();
    let mut seen = vec![false; n];
    let mut order = Vec::with_capacity(n);
    for u in 0..n {
        if !seen[u] {
            dfs_order(adjacency, u, &mut seen, &mut order);
        }
    }
    let rev = adjacency.transpose_view().to_csr();
    let mut labels = vec![usize::MAX; n];
    let mut cc = 0usize;
    while let Some(start) = order.pop() {
        if labels[start] != usize::MAX {
            continue;
        }
        let mut q = VecDeque::new();
        q.push_back(start);
        labels[start] = cc;
        while let Some(u) = q.pop_front() {
            if let Some(row) = rev.outer_view(u) {
                for &v in row.indices() {
                    if labels[v] == usize::MAX {
                        labels[v] = cc;
                        q.push_back(v);
                    }
                }
            }
        }
        cc += 1;
    }
    labels
}

/// Returns connected components.
pub fn get_connected_components(
    input_matrix: &CsMat<f64>,
    connection: &str,
    force_bipartite: bool,
) -> Result<Vec<usize>, StructureError> {
    if input_matrix.rows() == 0 || input_matrix.cols() == 0 {
        return Ok(Vec::new());
    }
    let (adjacency, _) = get_adjacency(
        MatrixInput::Sparse(input_matrix.to_owned()),
        true,
        force_bipartite,
        false,
        false,
    )
    .map_err(|_| StructureError::EmptyGraph)?;
    match connection {
        "weak" => Ok(weak_components(&adjacency)),
        "strong" => Ok(strong_components(&adjacency)),
        _ => Err(StructureError::InvalidConnection),
    }
}

/// Returns whether the input satisfies `is connected`.
pub fn is_connected(
    input_matrix: &CsMat<f64>,
    connection: &str,
    force_bipartite: bool,
) -> Result<bool, StructureError> {
    let labels = get_connected_components(input_matrix, connection, force_bipartite)?;
    if labels.is_empty() {
        return Ok(true);
    }
    Ok(labels.iter().all(|&x| x == labels[0]))
}

fn submatrix_rows_cols(input: &CsMat<f64>, rows: &[usize], cols: &[usize]) -> CsMat<f64> {
    let mut col_map = vec![usize::MAX; input.cols()];
    for (j_new, &j_old) in cols.iter().enumerate() {
        col_map[j_old] = j_new;
    }
    let mut tri = TriMat::<f64>::new((rows.len(), cols.len()));
    for (i_new, &i_old) in rows.iter().enumerate() {
        if let Some(row) = input.outer_view(i_old) {
            for (j_old, v) in row.iter() {
                let mapped = col_map[j_old];
                if mapped != usize::MAX {
                    tri.add_triplet(i_new, mapped, *v);
                }
            }
        }
    }
    tri.to_csr::<usize>()
}

/// Returns largest connected component.
pub fn get_largest_connected_component(
    input_matrix: &CsMat<f64>,
    connection: &str,
    force_bipartite: bool,
    return_index: bool,
) -> Result<(CsMat<f64>, Option<Vec<usize>>), StructureError> {
    let (adjacency, bipartite) = get_adjacency(
        MatrixInput::Sparse(input_matrix.to_owned()),
        true,
        force_bipartite,
        false,
        false,
    )
    .map_err(|_| StructureError::EmptyGraph)?;
    let labels = get_connected_components(&adjacency, connection, false)?;
    let n_labels = labels.iter().copied().max().unwrap_or(0) + 1;
    let mut counts = vec![0usize; n_labels];
    for &label in &labels {
        counts[label] += 1;
    }
    let largest_label = counts
        .iter()
        .enumerate()
        .max_by_key(|(_, c)| *c)
        .map(|(i, _)| i)
        .unwrap_or(0);

    if bipartite {
        let n_row = input_matrix.rows();
        let n_col = input_matrix.cols();
        let index_row: Vec<usize> = (0..n_row).filter(|&i| labels[i] == largest_label).collect();
        let index_col: Vec<usize> = (0..n_col)
            .filter(|&j| labels[n_row + j] == largest_label)
            .collect();
        let output = submatrix_rows_cols(input_matrix, &index_row, &index_col);
        let mut index = index_row.clone();
        index.extend(index_col.iter().copied());
        Ok((output, if return_index { Some(index) } else { None }))
    } else {
        let index: Vec<usize> = (0..input_matrix.rows())
            .filter(|&i| labels[i] == largest_label)
            .collect();
        let output = submatrix_rows_cols(input_matrix, &index, &index);
        Ok((output, if return_index { Some(index) } else { None }))
    }
}

/// Returns whether the input satisfies `is bipartite`.
pub fn is_bipartite(
    adjacency: &CsMat<f64>,
    return_biadjacency: bool,
) -> Result<BipartiteResult, StructureError> {
    if !is_symmetric(adjacency) {
        return Err(StructureError::NotUndirected);
    }
    let n = adjacency.rows();
    for i in 0..n.min(adjacency.cols()) {
        if adjacency.get(i, i).is_some() {
            return Ok(if return_biadjacency {
                BipartiteResult::Full(false, None, None, None)
            } else {
                BipartiteResult::Bool(false)
            });
        }
    }
    let mut coloring = vec![-1i8; n];
    for src in 0..n {
        if coloring[src] != -1 {
            continue;
        }
        let mut stack = vec![src];
        coloring[src] = 0;
        while let Some(node) = stack.pop() {
            if let Some(row) = adjacency.outer_view(node) {
                for &neighbor in row.indices() {
                    if coloring[neighbor] == -1 {
                        coloring[neighbor] = 1 - coloring[node];
                        stack.push(neighbor);
                    } else if coloring[neighbor] == coloring[node] {
                        return Ok(if return_biadjacency {
                            BipartiteResult::Full(false, None, None, None)
                        } else {
                            BipartiteResult::Bool(false)
                        });
                    }
                }
            }
        }
    }

    if return_biadjacency {
        let rows: Vec<usize> = (0..n).filter(|&i| coloring[i] == 0).collect();
        let cols: Vec<usize> = (0..n).filter(|&i| coloring[i] == 1).collect();
        let biadj = submatrix_rows_cols(adjacency, &rows, &cols);
        Ok(BipartiteResult::Full(
            true,
            Some(biadj),
            Some(rows),
            Some(cols),
        ))
    } else {
        Ok(BipartiteResult::Bool(true))
    }
}

#[cfg(test)]
mod tests {
    use ndarray::array;

    use super::*;
    use crate::utils::format::{bipartite2undirected, directed2undirected};

    fn dense_to_csr(dense: &ndarray::Array2<f64>) -> CsMat<f64> {
        let (r, c) = dense.dim();
        let mut tri = TriMat::<f64>::new((r, c));
        for i in 0..r {
            for j in 0..c {
                if dense[[i, j]] != 0.0 {
                    tri.add_triplet(i, j, dense[[i, j]]);
                }
            }
        }
        tri.to_csr::<usize>()
    }

    #[test]
    fn test_components_and_connected() {
        let digraph = dense_to_csr(&array![[0., 1., 0.], [0., 0., 1.], [0., 0., 0.]]);
        let labels = get_connected_components(&digraph, "weak", false).unwrap();
        assert_eq!(labels.len(), 3);
        assert!(is_connected(&digraph, "weak", false).unwrap());
        assert!(!is_connected(&digraph, "strong", false).unwrap());
    }

    #[test]
    fn test_largest_component() {
        let adjacency = dense_to_csr(&array![
            [0., 1., 0., 0.],
            [1., 0., 0., 0.],
            [0., 0., 0., 1.],
            [0., 0., 1., 0.]
        ]);
        let (largest, index) =
            get_largest_connected_component(&adjacency, "weak", false, true).unwrap();
        assert_eq!(largest.shape(), (2, 2));
        assert_eq!(index.unwrap().len(), 2);
    }

    #[test]
    fn test_is_bipartite() {
        let biadjacency = dense_to_csr(&array![[1., 0., 1.], [0., 1., 0.]]);
        let adjacency = bipartite2undirected(&biadjacency);
        let result = is_bipartite(&adjacency, false).unwrap();
        assert_eq!(result, BipartiteResult::Bool(true));

        let cyclic3 = directed2undirected(
            &dense_to_csr(&array![[0., 1., 0.], [0., 0., 1.], [1., 0., 0.]]),
            true,
        );
        let result = is_bipartite(&cyclic3, false).unwrap();
        assert_eq!(result, BipartiteResult::Bool(false));
    }
}
